pallets/unified-accounts/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+    },
+};
+use frame_system::{ensure_signed, pallet_prelude::*};
+use pallet_evm::AddressMapping;
+use sp_std::marker::PhantomData;
+
+pub use pallet::*;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+mod mock;
+mod tests;
+
+mod impls;
+pub use impls::*;
+
+type SignatureOf<T> = <<T as Config>::SignatureHelper as SignatureHelper>::Signature;
+
+/// Mapping between Native and EVM Addresses
+pub trait UnifiedAddressMapper<AccountId> {
+    /// Gets the account id associated with given evm address, if mapped else None.
+    fn to_account_id(evm_address: &EvmAddress) -> Option<AccountId>;
+    /// Gets the account id associated with given evm address.
+    /// If no mapping exists, then return the default evm address.
+    fn to_account_id_or_default(evm_address: &EvmAddress) -> AccountId;
+    /// Gets the default account id which is associated with given evm address.
+    fn to_default_account_id(evm_address: &EvmAddress) -> AccountId;
+
+    /// Gets the evm address associated with given account id, if mapped else None.
+    fn to_h160(account_id: &AccountId) -> Option<EvmAddress>;
+    /// Gets the evm address associated with given account id.
+    /// If no mapping exists, then return the default account id.
+    fn to_h160_or_default(account_id: &AccountId) -> EvmAddress;
+    /// Gets the default evm address which is associated with given account id.
+    fn to_default_h160(account_id: &AccountId) -> EvmAddress;
+}
+
+/// Signature verification scheme for proving address ownership
+pub trait SignatureHelper {
+    type AccountId;
+    type Address;
+    /// Signature type, ideally a 512-bit value for ECDSA signatures
+    type Signature: Parameter;
+
+    /// Build raw payload that user will sign (keccack hashed).
+    fn build_signing_payload(who: &Self::AccountId) -> [u8; 32];
+    /// Verify the provided signature against the given account.
+    fn verify_signature(who: &Self::AccountId, sig: &Self::Signature) -> Option<Self::Address>;
+}
+
+#[frame_support::pallet(dev_mode)]
+pub mod pallet {
+    use super::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The overarching event type
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        /// The Currency for managing evm address assets
+        type Currency: Mutate<Self::AccountId>;
+        /// Default evm address to account id conversion
+        type DefaultAddressMapping: AddressMapping<Self::AccountId>;
+        /// Default account id to evm address conversion
+        type DefaultAccountMapping: AccountMapping<Self::AccountId>;
+        /// The Signature verification implementation to use for checking claims
+        /// Note: the signature type defined by this will be used as parameter in pallet's extrinsic
+        type SignatureHelper: SignatureHelper<AccountId = Self::AccountId, Address = EvmAddress>;
```

Comment:
I don't think signature verification should be runtime configurable. Instead, the impl needs to be a solid built-in in `unified-account` pallet and offer no chance of being inconsistent between runtimes. `ChainId` should be configurable.